mindustry: build Arc from source

petabyteboy opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Currently we use binary shared libraries from maven. These include statically linked variants of SDL2 and glew from the author's build machine. To run mindustry on wayland without Xwayland it is necessary to link against a different glew library.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

nagy commented on pkgs/games/mindustry/default.nix:
```diff
@@ -93,20 +106,84 @@ let
 
+  glew-egl = glew.overrideAttrs (oldAttrs: {
+    pname = "glew-egl";
+    makeFlags = [ "SYSTEM=linux-egl" ];
```
Maybe this line should be:
```nix
makeFlags = oldAttrs.makeFlags ++ [ "SYSTEM=linux-egl" ];
```
This way, we pick up other `makeFlags` from the underlying derivation in the future. I know that there is already a `"SYSTEM=...` line in there, but as far as I know `make` will override those flags, so it could work. But please check for yourself, if this change makes sense.

petabyteboy commented:
This works now, but the whole file definitely needs a cleanup because it is currently _very_ messy.
There are way too many fixed-output derivation hashes that need to be redone on every update. We should automate that with an update script (think nix-universal-prefetch) and also add a check that ensures the isn't outdated (like buildRustPackage does with the cargo vendor hash).

petabyteboy commented:
![screenshot](https://user-images.githubusercontent.com/3250809/103975000-5ee3ac00-5173-11eb-9f66-79a6bebd59b1.png)

fgaz commented:
I wonder if this fixes the aarch64 build too

petabyteboy commented:
Partially: libsdl-arc64.so can be built, but libarcarm64.so, which is also used for sprite packing at build-time, is still missing. On x86_64-linux this is blob is downloaded from maven. All of these binary downloads that are not easily packagable all the way down are so ugly I can't even, but this is a generic Java problem.
I'm also wondering how this affects non-linux platforms, but it's difficult for me to test this.

petabyteboy commented:
https://github.com/Anuken/Mindustry/issues/1704#issuecomment-599093145

:frowning_face:

petabyteboy commented:
I can now build both libarc64.so and libsdl-arc64.so on x86_64-linux and inject them into the downloaded maven deps before starting the build. This should also make some of the LD_LIBRARY_PATH hacks unnecessary, since libarc64.so now has the correct rpath with alsaLib and so on.
However, porting this to aarch64 will be a bit more difficult: Upstream has some hard-coded compiler flags that can not work with aarch64.

fgaz commented:
That's unfortunate.

While you're at it then could you set meta.platforms to x86_64 and remove aarch64 from meta.broken, since it's unsupported upstream?

petabyteboy commented:
> While you're at it then could you set meta.platforms to x86_64 and remove aarch64 from meta.broken, since it's unsupported upstream?

Done.

---

Okay, so this has turned into a complete refactor:

- I didn't like how we had to build the whole project twice each time (once for the -deps), so I investigated a bit and found a solution to only prefetch all dependencies in the -deps derivation, without building anything.
- By accident I discovered that the Mindustry gradle project will discover an Arc checkout if it's placed in `../Arc` (probably mostly intended for development) and use that instead of attempting to download prebuilt versions. So I used this to combine the Arc and Mindustry builds into one derivation, going back to one -deps derivation as well.

This might have broken some more things on darwin, but since it didn't work before and it doesn't look like that problem will be solved soon, I would like to ignore that for now.

fgaz commented on pkgs/games/mindustry/default.nix:
```diff
@@ -25,12 +33,35 @@ let
+    cp ${stb.src}/stb_image.h Arc/arc-core/csrc/
+    cp -r ${soloud} Arc/arc-core/csrc/soloud
+    chmod -R u+w -- Arc
+    sourceRoot=Mindustry
```
Is there a reason why sourceRoot isn't at top level?

petabyteboy replied:
No specific reason, no

petabyteboy replied:
Changed to have the sourceRoot at top level.

fgaz commented on pkgs/games/mindustry/default.nix:
```diff
@@ -63,15 +91,26 @@ let
+    postPatch = cleanupMindustrySrc;
+
     nativeBuildInputs = [ gradle_6 perl ];
     # Here we build both the server and the client so we only have to specify
```
s/build/fetch the dependencies of/?

petabyteboy replied:
done.

fgaz commented on pkgs/games/mindustry/default.nix:
```diff
@@ -63,15 +91,26 @@ let
-      gradle --no-daemon server:dist -Pbuildversion=${buildVersion}
+    buildPhase = preBuild + ''
+      cat << EOF >> build.gradle
+      task resolveDependencies {
```
I wonder if upstream would accept a patch that adds this

petabyteboy replied:
Good idea, we'll find out: https://github.com/Anuken/Mindustry/pull/4302

fgaz commented on pkgs/games/mindustry/default.nix:
```diff
@@ -81,66 +120,76 @@ let
+    rm natives/natives-desktop/libs/libarc*.so
+    rm backends/backend-sdl/libs/linux64/libsdl-arc*.so
+
+    cd ../Mindustry
```
suggestion: pushd ../Arc in prePatch and popd here

petabyteboy replied:
I found an even better way: We can specify an `extraPrefix` as a parameter to `fetchpatch`, which will also make it possible to patch Arc and Mindustry at the same time.

fgaz commented on pkgs/games/mindustry/default.nix:
```diff
@@ -81,66 +120,76 @@ let
+  '' + cleanupMindustrySrc;
+
+  buildInputs = [ SDL2 glew alsaLib ];
+  nativeBuildInputs = [ pkg-config gradle_6 makeWrapper jdk ant copyDesktopItems ];
```
I like to have these on separate lines, so diffs are more granular

petabyteboy replied:
Done.

fgaz commented on pkgs/games/mindustry/default.nix:
```diff
@@ -81,66 +120,76 @@ let
     mkdir -p $out/bin
-    makeWrapper ${jre}/bin/java $out/bin/mindustry \
-      ${stdenv.lib.optionalString stdenv.isLinux "--prefix LD_LIBRARY_PATH : ${alsaLib}/lib"} \
+    makeWrapper ${jdk}/bin/java $out/bin/mindustry \
```
why jdk (later too)?

petabyteboy replied:
Maybe this should have been done in a seperate commit with an explanation, but there is a reason:
For modern java versions (>8) there is no JRE package anymore (the JDK is intended to be used for development and at runtime), so the `jre` attribute points to a very outdated version of Java.

petabyteboy replied:
If there was some specific reason for using `jre` with Java version 8, please unresolve this.

fgaz reviewed (approved):
>    I didn't like how we had to build the whole project twice each time (once for the -deps), so I investigated a bit and found a solution to only prefetch all dependencies in the -deps derivation, without building anything.

This is great! I also tried to do something like this some time ago but failed miserably...

>    By accident I discovered that the Mindustry gradle project will discover an Arc checkout if it's placed in ../Arc (probably mostly intended for development) and use that instead of attempting to download prebuilt versions. So I used this to combine the Arc and Mindustry builds into one derivation, going back to one -deps derivation as well.

Also good news! Only two hashes is definitely more manageable. Did you discover a way to find invalid hashes?

---

Thanks for the PR, +1 from me!

I left some minor comments here and there.

fgaz commented:
Conflict

petabyteboy commented:
> Conflict

is resolved

fgaz reviewed (approved):
Weird. Github still shows it as conflicting. Oh well.

fgaz commented on pkgs/games/mindustry/default.nix:
```diff
@@ -81,66 +132,77 @@ let
+    makeWrapper ${jdk}/bin/java $out/bin/mindustry \
       --add-flags "-jar $out/share/mindustry.jar"
     install -Dm644 core/assets/icons/icon_64.png $out/share/icons/hicolor/64x64/apps/mindustry.png
     install -Dm644 ${desktopItem}/share/applications/Mindustry.desktop $out/share/applications/Mindustry.desktop
```
wait I think you forgot to remove this since you use copyDesktopItems

petabyteboy replied:
Of course, you're right

Commit:
mindustry: build Arc from source

- reduce the buildPhase of the fixed-output dependencies derivation to
  only download all dependencies, instead of building the full thing.
- add wayland variant, which is linked against a different variant of
  the glew library to run natively using SDL_VIDEODRIVER=wayland
- use jdk (Java 15) instead of jre (Java 8) at runtime
- use new copyDesktopItems hook

petabyteboy commented:
Sorry I forgot to push earlier.

fgaz commented:
thank you! one binary less in nixpkgs :tada:

petabyteboy commented:
There will probably be a followup PR with some minor refinements and aarch64 support in the next few days (also planning to add aarch64 support upstream)
